File: Python/Sorting/Insertion-sort.py

# This code is contributed by Vivek Vijayan
# http://github.com/vivek-vijayan
# Sorting Head - start with 1
SortingHead = 1   

# Array list
array = []

while str(input("Add new value? (y/n) ")) == "y":
    array.append(input("List input  :"))


# Main Algorithm - Insertion Algorithm
while SortingHead != len(array):
    if(array[SortingHead] < array[SortingHead - 1]):
        # Swap the element if a[i] < a[i-1]
        array[SortingHead], array[SortingHead - 1] = array[SortingHead - 1], array[SortingHead]
        
        # Bring the header to before element as it is swapped. ignore if it swapped the first element as index is 0 
        if SortingHead != 1:
            SortingHead -= 1
    else:
        SortingHead += 1

# Printing the sorted array
print(array)

File: C++/Sorting/Insertion-sort.cpp
// This code is contributed by Vivek Vijayan
// http://github.com/vivek-vijayan

/*
    Insertion Sort 
    ---------------
    Average & Worst case - O(n^2)
*/

#include "iostream"
#include "vector"


int main(int argc, char const *argv[])
{
    std::vector <int> array;
    char option = 'y';
    int data;

    getdata :   std::cout << " Enter the element you want to enter : " << std::endl;
                std::cin >>  data;
                array.push_back(data);
                std::cout << "Add more element? (y/n) " << std::endl;
                std::cin >> option; 
                if (option=='y') goto getdata;
                std::cout << "Data recorded" << std::endl;
    
    // Main Algorithm - Insertion sort
    int SortingHead = 1, temp;
    while(SortingHead != array.size())
    {
        if(array[SortingHead] < array[SortingHead - 1]){
            temp = array[SortingHead];
            array[SortingHead] = array[SortingHead - 1];
            array[SortingHead - 1] = temp;

            if(SortingHead != 1){
                SortingHead--;
            }
        }
        else{
            SortingHead +=1;
        }
    }

    std::vector <int> :: iterator i;
    std::cout << "Sorted array : " << std::endl;
    for(i = array.begin(); i != array.end(); i++){
        std::cout << *i << std::endl;
    }
    return 0;
}
